File: tutorials/generate_prereq_practice/js/generate-prereq-practice.js
/* ================================
   CONFIGURATION & SETUP
   ================================ */

// Dependencies
const fs = require('fs');
const path = require('path');
const { parse } = require('csv-parse/sync');
const OpenAI = require('openai');
require('dotenv').config();

// Constants
const GENERATE_PRACTICE = true;
// Filter criteria for mathematics standards
const JURISDICTION = 'Multi-State';
const ACADEMIC_SUBJECT = 'Mathematics';
const TARGET_CODE = '6.NS.B.4';
// OpenAI configuration
const OPENAI_MODEL = 'gpt-4';
const OPENAI_TEMPERATURE = 0.7;

// Environment setup
const dataDir = process.env.KG_DATA_PATH;
if (!dataDir) {
  console.error('❌ KG_DATA_PATH environment variable is not set.');
  process.exit(1);
}

const openai = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY,
});

/* ================================
   HELPER FUNCTIONS
   ================================ */

function loadCSV(filename) {
  try {
    const content = fs.readFileSync(path.join(dataDir, filename), 'utf8');
    return parse(content, { columns: true, skip_empty_lines: true });
  } catch (error) {
    console.error(`❌ Error loading CSV file ${filename}: ${error.message}`);
    throw error;
  }
}


/* ================================
   STEP 1: LOAD DATA
   ================================ */
function loadData(aq) {
  /* Load CSV data files and build filtered dataset
   */

  const standardsFrameworkItems = aq.from(loadCSV('StandardsFrameworkItem.csv'));
  const learningComponents = aq.from(loadCSV('LearningComponent.csv'));
  const relationships = aq.from(loadCSV('Relationships.csv'));

  console.log('✅ Files loaded from KG CSV files');
  console.log({
    standardsFrameworkItems: standardsFrameworkItems.numRows(),
    learningComponents: learningComponents.numRows(),
    relationships: relationships.numRows()
  });

  // Filter for relevant StandardsFrameworkItems by jurisdiction and subject
  const relevantStandards = standardsFrameworkItems
    .params({ jurisdiction: JURISDICTION, academicSubject: ACADEMIC_SUBJECT })
    .filter(d => d.jurisdiction === jurisdiction && d.academicSubject === academicSubject);

  // Get array of relevant identifiers for filtering
  const relevantStandardIds = relevantStandards.array('caseIdentifierUUID');
  const relevantStandardSet = new Set(relevantStandardIds);

  // Filter relationships for buildsTowards and supports relationships
  const relevantRelationships = relationships
    .filter(aq.escape(d =>
      (d.relationshipType === 'buildsTowards' &&
        relevantStandardSet.has(d.sourceEntityValue) &&
        relevantStandardSet.has(d.targetEntityValue)) ||
      (d.relationshipType === 'supports' &&
        relevantStandardSet.has(d.targetEntityValue))
    ));

  // Get learning component IDs from supports relationships
  const supportRelationships = relevantRelationships
    .filter(d => d.relationshipType === 'supports');
  const linkedLearningComponentIds = supportRelationships.array('sourceEntityValue');
  const linkedLearningComponentSet = new Set(linkedLearningComponentIds);

  // Filter learning components by identifier
  const relevantLearningComponents = learningComponents
    .filter(aq.escape(d => linkedLearningComponentSet.has(d.identifier)));

  console.log('✅ Retrieved scoped graph:');
  console.log({
    standardsFrameworkItems: relevantStandards.numRows(),
    learningComponents: relevantLearningComponents.numRows(),
    relationships: relevantRelationships.numRows()
  });

  return {
    relevantStandards,
    relevantRelationships,
    relevantLearningComponents
  };
}

/* ================================
   STEP 2: QUERY PREREQUISITE DATA
   ================================ */
function getStandardAndPrerequisites(relevantStandards, relevantRelationships) {
  const targetStandardTable = relevantStandards
    .params({ targetCode: TARGET_CODE })
    .filter(d => d.statementCode === targetCode);

  if (targetStandardTable.numRows() === 0) {
    console.error(`❌ No StandardsFrameworkItem found for statementCode = "${TARGET_CODE}"`);
    return null;
  }

  const targetStandard = targetStandardTable.object();
  console.log(`✅ Found StandardsFrameworkItem for ${TARGET_CODE}:`)
  console.log({
    caseIdentifierUUID: targetStandard.caseIdentifierUUID,
    statementCode: targetStandard.statementCode,
    description: targetStandard.description
  });

  const prerequisiteLinks = relevantRelationships
    .params({ targetIdentifier: targetStandard.caseIdentifierUUID })
    .filter(d => d.relationshipType === 'buildsTowards' &&
      d.targetEntityValue === targetIdentifier);

  const prerequisiteStandards = prerequisiteLinks
    .join(relevantStandards, ['sourceEntityValue', 'caseIdentifierUUID'])
    .select('sourceEntityValue', 'statementCode', 'description_2')
    .rename({ sourceEntityValue: 'caseIdentifierUUID', description_2: 'standardDescription' });

  console.log(`✅ Found ${prerequisiteStandards.numRows()} prerequisite(s) for ${targetStandard.statementCode}:`);
  console.log(prerequisiteStandards.objects());

  return { targetStandard, prerequisiteStandards };
}

function getLearningComponentsForPrerequisites(prerequisiteStandards, relevantRelationships, relevantLearningComponents) {
  const prerequisiteLearningComponents = prerequisiteStandards
    .join(relevantRelationships, ['caseIdentifierUUID', 'targetEntityValue'])
    .params({ supportsType: 'supports' })
    .filter(d => d.relationshipType === supportsType)
    .join(relevantLearningComponents, ['sourceEntityValue', 'identifier'])
    .select('caseIdentifierUUID', 'statementCode', 'standardDescription', 'description_2')
    .rename({ description_2: 'learningComponentDescription' });

  console.log(`✅ Found ${prerequisiteLearningComponents.numRows()} supporting learning components for prerequisites:`);
  console.log(prerequisiteLearningComponents.objects());

  return prerequisiteLearningComponents;
}

function queryPrerequisiteData(aq, relevantStandards, relevantRelationships, relevantLearningComponents) {
  /* Analyze prerequisite relationships for the target standard
   * This step identifies prerequisites and supporting learning components
   * 
   * SQL:    WITH target AS (
   *           SELECT "caseIdentifierUUID"
   *           FROM standards_framework_item
   *           WHERE "statementCode" = '6.NS.B.4'
   *         ),
   *         prerequisite_standards AS (
   *           SELECT
   *             sfi."caseIdentifierUUID",
   *             sfi."statementCode",
   *             sfi."description"
   *           FROM standards_framework_item sfi
   *           JOIN relationships r
   *             ON sfi."caseIdentifierUUID" = r."sourceEntityValue"
   *           JOIN target t
   *             ON r."targetEntityValue" = t."caseIdentifierUUID"
   *           WHERE r."relationshipType" = 'buildsTowards'
   *         )
   *         SELECT
   *           ps."caseIdentifierUUID",
   *           ps."statementCode",
   *           ps."description",
   *           lc."description"
   *         FROM prerequisite_standards ps
   *         JOIN relationships r
   *           ON ps."caseIdentifierUUID" = r."targetEntityValue"
   *         JOIN learning_component lc
   *           ON r."sourceEntityValue" = lc."identifier"
   *         WHERE r."relationshipType" = 'supports';
   * 
   * Cypher: MATCH (target:StandardsFrameworkItem {statementCode: '6.NS.B.4'})
   *         MATCH (prereq:StandardsFrameworkItem)-[:buildsTowards]->(target)
   *         MATCH (lc:LearningComponent)-[:supports]->(prereq)
   *         RETURN prereq.caseIdentifierUUID, prereq.statementCode, prereq.description, 
   *                lc.description
   */

  const standardAndPrereqData = getStandardAndPrerequisites(relevantStandards, relevantRelationships);
  if (!standardAndPrereqData) {
    return null;
  }

  const { targetStandard, prerequisiteStandards } = standardAndPrereqData;
  const prerequisiteLearningComponents = getLearningComponentsForPrerequisites(prerequisiteStandards, relevantRelationships, relevantLearningComponents);

  return { targetStandard, prerequisiteLearningComponents };
}

/* ================================
   STEP 3: GENERATE PRACTICE
   ================================ */
function packageContextData(targetStandard, prerequisiteLearningComponents) {
  /* Package the standards and learning components data for text generation
   * This creates a structured context that can be used for generating practice questions
   */

  // Convert dataframe to context format for LLM
  const allRows = prerequisiteLearningComponents.objects();
  const standardsMap = new Map();

  // Group learning components by standard for context
  for (const row of allRows) {
    if (!standardsMap.has(row.caseIdentifierUUID)) {
      standardsMap.set(row.caseIdentifierUUID, {
        statementCode: row.statementCode,
        description: row.standardDescription || '(no statement)',
        supportingLearningComponents: []
      });
    }

    standardsMap.get(row.caseIdentifierUUID).supportingLearningComponents.push({
      description: row.learningComponentDescription || '(no description)'
    });
  }

  const fullStandardsContext = {
    targetStandard: {
      statementCode: targetStandard.statementCode,
      description: targetStandard.description || '(no statement)'
    },
    prereqStandards: Array.from(standardsMap.values())
  };

  console.log(`✅ Packaged full standards context for text generation`);
  console.log(JSON.stringify(fullStandardsContext, null, 2));
  return fullStandardsContext;
}

function generatePracticeData(fullStandardsContext) {
  /* Generate practice questions using OpenAI API
   * This creates educational content based on prerequisite data
   */
  return async function generatePractice() {
    console.log(`🔄 Generating practice questions for ${fullStandardsContext.targetStandard.statementCode}...`);

    try {
      // Build prompt inline
      let prerequisiteText = '';
      for (const prereq of fullStandardsContext.prereqStandards) {
        prerequisiteText += `- ${prereq.statementCode}: ${prereq.description}\n`;
        prerequisiteText += '  Supporting Learning Components:\n';
        for (const lc of prereq.supportingLearningComponents) {
          prerequisiteText += `    • ${lc.description}\n`;
        }
      }

      const prompt = `You are a math tutor helping middle school students. Based on the following information, generate 3 practice questions for the target standard. Questions should help reinforce the key concept and build on prerequisite knowledge.

Target Standard:
- ${fullStandardsContext.targetStandard.statementCode}: ${fullStandardsContext.targetStandard.description}

Prerequisite Standards & Supporting Learning Components:
${prerequisiteText}`;

      const response = await openai.chat.completions.create({
        model: OPENAI_MODEL,
        messages: [
          { role: 'system', content: 'You are an expert middle school math tutor.' },
          { role: 'user', content: prompt }
        ],
        temperature: OPENAI_TEMPERATURE
      });

      const practiceQuestions = response.choices[0].message.content.trim();

      console.log(`✅ Generated practice questions:\n`);
      console.log(practiceQuestions);

      return {
        aiGenerated: practiceQuestions,
        targetStandard: fullStandardsContext.targetStandard.statementCode,
        prerequisiteCount: fullStandardsContext.prereqStandards.length
      };
    } catch (err) {
      console.error('❌ Error generating practice questions:', err.message);
      throw err;
    }
  };
}


async function main() {
  console.log('\n=== GENERATE PREREQUISITE PRACTICE TUTORIAL ===\n');

  console.log('🔄 Step 1: Loading data...');
  const aq = await import('arquero');
  const { relevantStandards, relevantRelationships, relevantLearningComponents } = loadData(aq);

  console.log('\n🔄 Step 2: Querying prerequisite data...');
  const prerequisiteData = queryPrerequisiteData(aq, relevantStandards, relevantRelationships, relevantLearningComponents);

  if (!prerequisiteData) {
    console.error('❌ Failed to find prerequisite data');
    return;
  }

  const { targetStandard, prerequisiteLearningComponents } = prerequisiteData;

  console.log('\n🔄 Step 3: Generating practice...');
  const fullStandardsContext = packageContextData(targetStandard, prerequisiteLearningComponents);
  const generatePractice = generatePracticeData(fullStandardsContext);
  if (GENERATE_PRACTICE) {
    await generatePractice();
  } else {
    console.log('🚫 Practice question generation disabled');
  }
}

main().catch(console.error);

File: tutorials/generate_prereq_practice/python/generate_prereq_practice.py
#!/usr/bin/env python3
"""
================================
CONFIGURATION & SETUP
================================
"""

# Dependencies
import os
import sys
import json
import pandas as pd
from pathlib import Path
from dotenv import load_dotenv
from openai import OpenAI

# Load environment variables
load_dotenv()

# Domain Constants
GENERATE_PRACTICE = True
# Filter criteria for mathematics standards
JURISDICTION = 'Multi-State'
ACADEMIC_SUBJECT = 'Mathematics'
TARGET_CODE = '6.NS.B.4'
# OpenAI configuration
OPENAI_MODEL = 'gpt-4'
OPENAI_TEMPERATURE = 0.7

# Environment Setup
data_dir = os.getenv('KG_DATA_PATH')
if not data_dir:
    print('❌ KG_DATA_PATH environment variable is not set.')
    sys.exit(1)

data_path = Path(data_dir)

openai_client = OpenAI(
    api_key=os.getenv('OPENAI_API_KEY')
)


"""
================================
HELPER FUNCTIONS
================================
"""

def load_csv(filename):
    """Load and parse CSV file from data directory"""
    try:
        file_path = data_path / filename
        return pd.read_csv(file_path)
    except Exception as error:
        print(f'❌ Error loading CSV file {filename}: {str(error)}')
        raise error


"""
================================
STEP 1: LOAD DATA
================================
"""

def load_data():
    """Load CSV data files and build filtered dataset"""
    
    standards_framework_items = load_csv('StandardsFrameworkItem.csv')
    learning_components = load_csv('LearningComponent.csv')
    relationships = load_csv('Relationships.csv')

    print('✅ Files loaded from KG CSV files')
    print(f'  Standards Framework Items: {len(standards_framework_items)}')
    print(f'  Learning Components: {len(learning_components)}')
    print(f'  Relationships: {len(relationships)}')

    # Filter for relevant StandardsFrameworkItems by jurisdiction and subject
    relevant_standards = standards_framework_items[
        (standards_framework_items['jurisdiction'] == JURISDICTION) &
        (standards_framework_items['academicSubject'] == ACADEMIC_SUBJECT)
    ]

    # Get set of relevant identifiers for filtering
    relevant_standard_ids = set(relevant_standards['caseIdentifierUUID'])

    # Filter relationships for buildsTowards and supports relationships
    relevant_relationships = relationships[
        ((relationships['relationshipType'] == 'buildsTowards') &
         (relationships['sourceEntityValue'].isin(relevant_standard_ids)) &
         (relationships['targetEntityValue'].isin(relevant_standard_ids))) |
        ((relationships['relationshipType'] == 'supports') &
         (relationships['targetEntityValue'].isin(relevant_standard_ids)))
    ]

    # Get learning component IDs from supports relationships
    support_relationships = relevant_relationships[
        relevant_relationships['relationshipType'] == 'supports'
    ]
    linked_learning_component_ids = set(support_relationships['sourceEntityValue'])

    # Filter learning components by identifier
    relevant_learning_components = learning_components[
        learning_components['identifier'].isin(linked_learning_component_ids)
    ]

    print('✅ Retrieved scoped graph:')
    print(f'  Standards Framework Items: {len(relevant_standards)}')
    print(f'  Learning Components: {len(relevant_learning_components)}')
    print(f'  Relationships: {len(relevant_relationships)}')

    return {
        'relevant_standards': relevant_standards,
        'relevant_relationships': relevant_relationships,
        'relevant_learning_components': relevant_learning_components
    }


"""
================================
STEP 2: QUERY PREREQUISITE DATA
================================
"""

def get_standard_and_prerequisites(relevant_standards, relevant_relationships):
    target_standard_df = relevant_standards[
        relevant_standards['statementCode'] == TARGET_CODE
    ]

    if len(target_standard_df) == 0:
        print(f'❌ No StandardsFrameworkItem found for statementCode = "{TARGET_CODE}"')
        return None

    target_standard = target_standard_df.iloc[0]
    print(f'✅ Found StandardsFrameworkItem for "{TARGET_CODE}":')
    print(f'  UUID: {target_standard["caseIdentifierUUID"]}')
    print(f'  Statement Code: {target_standard["statementCode"]}')
    print(f'  Description: {target_standard["description"]}')

    prerequisite_links = relevant_relationships[
        (relevant_relationships['relationshipType'] == 'buildsTowards') &
        (relevant_relationships['targetEntityValue'] == target_standard['caseIdentifierUUID'])
    ]

    prerequisite_standards = prerequisite_links.merge(
        relevant_standards,
        left_on='sourceEntityValue',
        right_on='caseIdentifierUUID',
        suffixes=('_rel', '_std')
    )[['sourceEntityValue', 'statementCode', 'description_std']].rename(columns={
        'sourceEntityValue': 'caseIdentifierUUID',
        'description_std': 'standardDescription'
    })

    print(f'✅ Found {len(prerequisite_standards)} prerequisite(s) for {target_standard["statementCode"]}:')
    for i, (_, std) in enumerate(prerequisite_standards.iterrows(), 1):
        print(f'  {i}. {std["statementCode"]}: {std["standardDescription"][:100]}...')

    return {'target_standard': target_standard, 'prerequisite_standards': prerequisite_standards}


def get_learning_components_for_prerequisites(prerequisite_standards, relevant_relationships, relevant_learning_components):
    prerequisite_learning_components = prerequisite_standards.merge(
        relevant_relationships,
        left_on='caseIdentifierUUID',
        right_on='targetEntityValue',
        suffixes=('_std', '_rel')
    )
    
    prerequisite_learning_components = prerequisite_learning_components[
        prerequisite_learning_components['relationshipType'] == 'supports'
    ].merge(
        relevant_learning_components,
        left_on='sourceEntityValue',
        right_on='identifier',
        suffixes=('_rel', '_lc')
    )
    
    # Select the correct columns - after merge with suffixes, learning component description becomes description_lc
    prerequisite_learning_components = prerequisite_learning_components[
        ['caseIdentifierUUID', 'statementCode', 'standardDescription', 'description_lc']
    ].rename(columns={
        'description_lc': 'learningComponentDescription'
    })

    print(f'✅ Found {len(prerequisite_learning_components)} supporting learning components for prerequisites:')
    for i, (_, lc) in enumerate(prerequisite_learning_components.iterrows(), 1):
        print(f'  {i}. {lc["learningComponentDescription"][:100]}...')

    return prerequisite_learning_components


def query_prerequisite_data(relevant_standards, relevant_relationships, relevant_learning_components):
    """
    Analyze prerequisite relationships for the target standard
    This step identifies prerequisites and supporting learning components
    
    SQL:    WITH target AS (
              SELECT "caseIdentifierUUID"
              FROM standards_framework_item
              WHERE "statementCode" = '6.NS.B.4'
            ),
            prerequisite_standards AS (
              SELECT
                sfi."caseIdentifierUUID",
                sfi."statementCode",
                sfi."description"
              FROM standards_framework_item sfi
              JOIN relationships r
                ON sfi."caseIdentifierUUID" = r."sourceEntityValue"
              JOIN target t
                ON r."targetEntityValue" = t."caseIdentifierUUID"
              WHERE r."relationshipType" = 'buildsTowards'
            )
            SELECT
              ps."caseIdentifierUUID",
              ps."statementCode",
              ps."description",
              lc."description"
            FROM prerequisite_standards ps
            JOIN relationships r
              ON ps."caseIdentifierUUID" = r."targetEntityValue"
            JOIN learning_component lc
              ON r."sourceEntityValue" = lc."identifier"
            WHERE r."relationshipType" = 'supports';
    
    Cypher: MATCH (target:StandardsFrameworkItem {statementCode: '6.NS.B.4'})
            MATCH (prereq:StandardsFrameworkItem)-[:buildsTowards]->(target)
            MATCH (lc:LearningComponent)-[:supports]->(prereq)
            RETURN prereq.caseIdentifierUUID, prereq.statementCode, prereq.description, 
                   lc.description
    """

    standard_and_prereq_data = get_standard_and_prerequisites(relevant_standards, relevant_relationships)
    if not standard_and_prereq_data:
        return None

    target_standard = standard_and_prereq_data['target_standard']
    prerequisite_standards = standard_and_prereq_data['prerequisite_standards']
    prerequisite_learning_components = get_learning_components_for_prerequisites(prerequisite_standards, relevant_relationships, relevant_learning_components)

    return {'target_standard': target_standard, 'prerequisite_learning_components': prerequisite_learning_components}


"""
================================
STEP 3: GENERATE PRACTICE
================================
"""

def package_context_data(target_standard, prerequisite_learning_components):
    """
    Package the standards and learning components data for text generation
    This creates a structured context that can be used for generating practice questions
    """

    # Convert dataframe to context format for LLM
    all_rows = prerequisite_learning_components.to_dict('records')
    standards_map = {}

    # Group learning components by standard for context
    for row in all_rows:
        case_id = row['caseIdentifierUUID']
        if case_id not in standards_map:
            standards_map[case_id] = {
                'statementCode': row['statementCode'],
                'description': row['standardDescription'] or '(no statement)',
                'supportingLearningComponents': []
            }

        standards_map[case_id]['supportingLearningComponents'].append({
            'description': row['learningComponentDescription'] or '(no description)'
        })

    full_standards_context = {
        'targetStandard': {
            'statementCode': target_standard['statementCode'],
            'description': target_standard['description'] or '(no statement)'
        },
        'prereqStandards': list(standards_map.values())
    }

    print('✅ Packaged full standards context for text generation')
    print(json.dumps(full_standards_context, indent=2))
    return full_standards_context


def generate_practice_data(full_standards_context):
    """Generate practice questions using OpenAI API
    This creates educational content based on prerequisite data
    """
    def generate_practice():
        print(f'🔄 Generating practice questions for {full_standards_context["targetStandard"]["statementCode"]}...')

        try:
            # Build prompt inline
            prerequisite_text = ''
            for prereq in full_standards_context['prereqStandards']:
                prerequisite_text += f'- {prereq["statementCode"]}: {prereq["description"]}\n'
                prerequisite_text += '  Supporting Learning Components:\n'
                for lc in prereq['supportingLearningComponents']:
                    prerequisite_text += f'    • {lc["description"]}\n'

            prompt = f"""You are a math tutor helping middle school students. Based on the following information, generate 3 practice questions for the target standard. Questions should help reinforce the key concept and build on prerequisite knowledge.

Target Standard:
- {full_standards_context["targetStandard"]["statementCode"]}: {full_standards_context["targetStandard"]["description"]}

Prerequisite Standards & Supporting Learning Components:
{prerequisite_text}"""

            response = openai_client.chat.completions.create(
                model=OPENAI_MODEL,
                messages=[
                    {'role': 'system', 'content': 'You are an expert middle school math tutor.'},
                    {'role': 'user', 'content': prompt}
                ],
                temperature=OPENAI_TEMPERATURE
            )

            practice_questions = response.choices[0].message.content.strip()

            print('✅ Generated practice questions:\n')
            print(practice_questions)

            return {
                'aiGenerated': practice_questions,
                'targetStandard': full_standards_context['targetStandard']['statementCode'],
                'prerequisiteCount': len(full_standards_context['prereqStandards'])
            }
        except Exception as err:
            print(f'❌ Error generating practice questions: {str(err)}')
            raise err

    return generate_practice


"""
================================
MAIN EXECUTION
================================
"""

def main():
    """Main execution function - orchestrates all tutorial steps"""
    print('\n=== GENERATE PREREQUISITE PRACTICE TUTORIAL ===\n')

    print('🔄 Step 1: Loading data...')
    data = load_data()
    relevant_standards = data['relevant_standards']
    relevant_relationships = data['relevant_relationships']
    relevant_learning_components = data['relevant_learning_components']

    print('\n🔄 Step 2: Querying prerequisite data...')
    prerequisite_data = query_prerequisite_data(relevant_standards, relevant_relationships, relevant_learning_components)

    if not prerequisite_data:
        print('❌ Failed to find prerequisite data')
        return

    target_standard = prerequisite_data['target_standard']
    prerequisite_learning_components = prerequisite_data['prerequisite_learning_components']

    print('\n🔄 Step 3: Generating practice...')
    full_standards_context = package_context_data(target_standard, prerequisite_learning_components)
    generate_practice = generate_practice_data(full_standards_context)
    if GENERATE_PRACTICE:
        generate_practice()
    else:
        print('🚫 Practice question generation disabled')


if __name__ == '__main__':
    main()